Enable multiplatform configurations

natario1 opened the pull request:
This PR introduces sourceSet specific gradle configurations for multiplatform users, so that ksp processors can be applied selectively to only one target, for example. Relevant discussion: https://kotlinlang.slack.com/archives/C013BA8EQSE/p1630359069041200

#### Before PR
- Root configuration `ksp` incorrectly? applied to everything
- Source set configurations `ksp<Set>` available only in single-platform projects (like kotlin-jvm)
- Android treated as a special case, creating `ksp<Set>` configurations from AGP `AndroidSourceSet`s, but without dedicated logic when picking up the correct configurations to apply to a given compilation.

#### After PR
- <strike>Root configuration `ksp` is applied to main sets of all targets - does not apply to test compilations anymore</strike> 
- **Single-platform projects** (like kotlin-jvm): no changes, one configuration per set named `ksp<Set>`. Use `ksp` for the main set, and `kspTest` for the test set.
- **Multi-platform projects** (kotlin-multiplatform): just like single-platform. Note though that multiplatform source sets are named after the target, so the configuration name will also include it. For example for a JVM target, we have `kspJvm` and `kspJvmTest`. **This also means that the `ksp` configuration will not be available. See discussion here: https://github.com/google/ksp/pull/609#discussion_r714226247 for why.** This is a breaking change for multiplatform users - I tried to mitigate the impact of this change by throwing a clear error.
- Android still treated as before, but in multi-platform we add the target name prefix. If target is named `android`, configurations are named `kspAndroid<AndroidSourceSet>`. This is needed to avoid conflicts with other targets.

Note that I moved configuration logic to a separate `KspConfigurations` class for readability, and also fixed a few issues when creating the `ksp<>ProcessorClasspath` configuration.

gavra0 commented on gradle-plugin/src/test/kotlin/com/google/devtools/ksp/gradle/SourceSetConfigurationsTest.kt:
```diff
@@ -72,6 +72,47 @@ class SourceSetConfigurationsTest {
     }
 
+    @Test
+    fun configurationsForMultiplatformApp() {
```
Can you please extend the test to check that final configurations "_processorClasspath" are correctly constructed? E.g having something like this in the build file:
```
configurations.filter { "processorClasspath" }.forEach {
    if (it.name = "kspJvmProcessorClasspath") it.extendsFrom == ["main"]
 }
 ```

gavra0 commented on gradle-plugin/src/test/kotlin/com/google/devtools/ksp/gradle/SourceSetConfigurationsTest.kt:
```diff
@@ -72,6 +72,47 @@ class SourceSetConfigurationsTest {
+    @Test
+    fun configurationsForMultiplatformApp() {
+        testRule.setupAppAsMultiplatformApp(
+            "jvm { }",
```
nit: Can you pass a multi-line string instead?

natario1 commented on gradle-plugin/src/main/kotlin/com/google/devtools/ksp/gradle/KspConfigurations.kt:
```diff
@@ -0,0 +1,136 @@
+        } else {
+            // maybeCreate to be future-proof, but we should never have a duplicate with current logic
+            project.configurations.maybeCreate(configName).apply {
+                if (parent != null) extendsFrom(parent)
```
This is needed for ksp configurations that are not related to a source set. Currently there's one case where this happens - the root `ksp` configuration in a multiplatform project. But I think there might be more in the future, for example it wouldn't be bad to add `kspTest` in multiplatform and make it a parent of all test sets.

gavra0 replied:
All configurations should be associated to a source set, so that being said we probably do not want `ksp` in multiplatform project, it should always be `ksp<Target><SourceSetName>` or `ksp<Target>` (for the main source set of a target).

natario1 replied:
I can remove `ksp` for multiplatform but it's a breaking change, a very breaking one I would say 😄 it'd break all MP tests for example. What do you think?

gavra0 replied:
It is definitely a breaking change, but the current situation creates more confusion than benefit.

Adding multiple artifacts to `ksp` configuration (across source sets) impacts all of them. In contrast, `implementation` under MPP `sourceSets.jvmMain.dependencies` actually maps to configuration `implementationJvm` so there are no conflicts.

natario1 replied:
That was a feature in my mind. But I get your point, I will remove it. In the future I think we could add `KotlinDependenciesHandler.ksp()` extensions to be used in KMP just like implementation in your example - mapping to the appropriate `ksp<SourceSet>` configuration.

gavra0 commented on gradle-plugin/src/main/kotlin/com/google/devtools/ksp/gradle/KspConfigurations.kt:
```diff
@@ -0,0 +1,136 @@
+            parent
+        } else {
+            // maybeCreate to be future-proof, but we should never have a duplicate with current logic
+            project.configurations.maybeCreate(configName).apply {
```
Can you please add the following properties to the configuration:
- `description` e.g `KSP dependencies for jvmTest source set`
- set `setVisible(false)`
- set `canBeResolved = false` (as we should never resolve these directly)
- set `isCanBeConsumed = false` (to make sure this configuration is not used for publishing)

natario1 commented:
@gavra0 I think I addressed all your comments. I added this

https://github.com/google/ksp/blob/efed4a3cca7cc04eb57552b3fcf060e45e0bebff/gradle-plugin/src/main/kotlin/com/google/devtools/ksp/gradle/KspConfigurations.kt#L61-L68

to try to mitigate the impact of the breaking change. Let me know what you think.

gavra0 reviewed (commented):
Thank you covering all scenarios with tests!

natario1 commented on gradle-plugin/src/main/kotlin/com/google/devtools/ksp/gradle/KspConfigurations.kt:
```diff
@@ -0,0 +1,129 @@
+    }
+
+    // Store all ksp configurations for quick retrieval.
+    private val kotlinConfigurations = mutableMapOf<KotlinSourceSet, Configuration>()
+    private val androidConfigurations = mutableMapOf<String, Configuration>()
```
I'm not sure how to do this without complicating the logic. I'm using string for Android to keep all AGP types within `AndroidPluginIntegration` class (see docs of the class, AGP is `compileOnly`).

If we use the set name as a key for both Android and Kotlin, there is risk of conflict in KMP. `Named` would be a very good common type, but AGP SourceProvider does not extend it. I should either use `Any` or string + mapper functions, but that's even harder because the set alone is not enough to compute the key, need to know the target and compilation and whether it's the default one

gavra0 replied:
The reason why I'd like to remove this is to have as few as possible Android-specific things in KSP, as it makes maintenance much easier. We are a bit far from the end-goal though :)

Configuration names must be unique, and we already have logic in `decorateKotlinTarget(target: KotlinTarget)` that knows how to compute such name for `KotlinPlatformType.androidJvm` and `KotlinCompilation`. I think we can extract the name-computing logic from that method, and have it reused during configuration creation and then during configuration fetching. WDYT?

natario1 replied:
@gavra0 do we really need a `String > Configuration` map then? If key is the ksp configuration name, `project.configurations` could do the job already.

gavra0 replied:
Totally correct, we can drop the map. The less code, the better :)

natario1 replied:
Done

ting-yuan commented on gradle-plugin/src/main/kotlin/com/google/devtools/ksp/gradle/KspConfigurations.kt:
```diff
@@ -0,0 +1,129 @@
+                // Adding multiplatform configuration removed support for the root ksp configuration.
+                // Try to make this breaking change less breaking by adding a clear error.
+                project.configurations.create("ksp").dependencies.all {
+                    throw InvalidUserCodeException(
```
Pardon me if I missed any discussion in the review thread, but I'm wondering why taking this away. Isn't an apply-to-all configuration (except `kspTest`) useful?

natario1 replied:
It was discussed here https://github.com/google/ksp/pull/609#discussion_r714226247 I don't have a strong opinion FWIW, up to you guys. To summarize I think Ivan point was that in KMP if you do

```kotlin
kotlin {
    jvm()
    sourceSets.jvmMain {
        dependencies { this: KotlinDependenciesHandler ->
            implementation("foo")
            project.dependencies.add("ksp", "bar")
        }
    }
}
```

then `foo` is jvm only, while `bar` will be applied to all targets, and this is confusing. I think we could add `KotlinDependenciesHandler.ksp()` DSL extensions so that people just write `ksp(bar)`, but I'm not sure about the implementation (need to cast KDH to some internal Kotlin plugin type) and also didn't want to make this PR bigger.

gavra0 replied:
To echo Mattia's point, I think that having a global `ksp` configuration is quite error prone. If there is a need for a configuration that adds to all platforms, it should probably be named `kspAllPlatforms` or something like that.

ting-yuan replied:
Got it, thanks. I'm worrying about the breakage 😢 We are planning for 1.0.1 release and this PR is definitely the most important one to include. I'm wondering whether we should keep the universal `ksp` for now, and announce that we are going to remove it in a later release? So that people are less surprised and have a chance to migrate. As a mitigation, I can polish the doc, add an example of KMP and tell people to use `ksp<SourceSet>` instead of `ksp`.

What do you think?

gavra0 replied:
Agree, it does seem like the current approach may be a bit too aggressive. This migration can be more smooth by doing it in 3 stages:
- keep `ksp` and have other configurations extend from it but add warning that folks should migrate to source-set specific configurations
- keep `ksp` configuration, but make it a no-op, and tell users to migrate
- remove `ksp` configuration fully

gavra0 reviewed (approved):
This looks good to me, we can merge this as-is, and add back the `ksp` configuration in a new PR.

ting-yuan commented:
Thanks for the patch and review!
